host/app/lib/schema-util.ts

```diff
@@ -0,0 +1,188 @@
+      offset = match.end.index! + match.end[0].length;
+    }
+    output.push(this.src.slice(offset));
+    return output.join('');
+  }
+
+  // the semantic phase is async
+  private async semanticPhase() {
+    for (let card of this.possibleCards) {
+      if (await this.isCardReference(card.super)) {
+        let fields: Map<string, FieldDefinition> = new Map();
+        for (let [fieldName, field] of card.possibleFields) {
+          let fieldType = await this.getFieldType(field);
+          if (!fieldType) {
+            continue;
+          }
+          if (!(await this.isCardReference(field.card))) {
+            continue;
+          }
+          fields.set(fieldName, {
+            type: fieldType,
+            card: field.card,
+          });
+        }
+        this.cards.push(new CardDefinition(card, fields));
+      }
+    }
+  }
+
+  private async isCardReference(
+    possibleCardRef: CardReference
+  ): Promise<boolean> {
+    switch (possibleCardRef.type) {
+      case 'external': {
+        let mod = await this.inspector.resolveModule(possibleCardRef.module);
+        let superClass = mod[possibleCardRef.name];
+        return typeof superClass === 'function' && 'baseCard' in superClass;
+      }
+      case 'internal':
+        return await this.isCardReference(
+          this.possibleCards[possibleCardRef.classIndex].super
+        );
+      default:
+        throw assertNever(possibleCardRef);
+    }
+  }
+
+  private async getFieldType(
+    possibleField: PossibleField
+  ): Promise<FieldType | undefined> {
+    let decoratorMod = await this.inspector.resolveModule(
+      possibleField.decorator.module
+    );
+    if (!(fieldDecorator in decoratorMod[possibleField.decorator.name])) {
+      return undefined;
+    }
+    let fieldTypeMod = await this.inspector.resolveModule(
+      possibleField.type.module
+    );
+
+    if (!(await this.isCardReference(possibleField.card))) {
+      return undefined;
+    }
+    let type = fieldTypeMod[possibleField.type.name][fieldType];
+    if (!isFieldType(type)) {
+      return undefined;
+    }
+
+    return type;
+  }
+
+  cards: CardDefinition[] = [];
+}
+
+export class CardDefinition {
+  constructor(
+    private cardClass: PossibleCardClass,
+    private fieldDefinitions: Map<string, FieldDefinition>
+  ) {}
+
+  getField(name: string) {
+    return this.fieldDefinitions.get(name);
+  }
+  get localName() {
+    return this.cardClass.localName;
+  }
+  get exportedAs() {
+    return this.cardClass.exportedAs;
+  }
+  get fields() {
+    return [...this.fieldDefinitions];
+  }
+}
+
+export interface FieldDefinition {
+  card: CardReference;
+  type: FieldType;
+}
+
+export class CardInspector {
+  readonly resolveModule: (specifier: string) => Promise<Record<string, any>>;
+  // this is intentionally manipulated by the outside in order to set the
+  // current path for relative imports--seems like not such a great
+  // abstraction...
+  currentPath: string;
```

Comment:
another place where we are setting up the CardInspector to be manipulated from the outside--feels not good.